[로또] 쉐이크반 미션 제출합니다.

vsh123 opened the pull request:
안녕하세요!

로또게임 3단계까지 구현한 내용을 제출합니다!
로또 자동생성에 대해서는 Collections.shuffle이라는 메소드를 추천해주셨는데,
1~45의 숫자들을 Map에 저장하고 있어서 Random을 사용하였습니다.

List가 아닌 Map을 사용한 이유는 List를 사용할 경우 사용자가 요청한 index에 -1한 위치의 값을 리턴해야 하는데 
이 때 -1을 하는 것보다 Map을 사용하여 1~45사이의 키를 들고 있고 요청한 index에 있는 value를 리턴하는 것이
조금 더 깔끔하다고 생각하여 구현하였습니다.

혹시 위 내용에 대해서 제가 조금 더 고려해야할 점 있으면 피드백 부탁드리겠습니다!
그 외 다른 부분에서도 고쳐야할 점이 있다면 많이많이 피드백 주시면 감사하겠습니다~~ :)

감사합니다!

vsh123 committed:
docs : 1단계 기능 구현 목록 작성

vsh123 committed:
feat : LottoNumber 객체 생성

 1. 1~45개의 LottoNumber인스턴스 객체를 가지고 있는 클래스 생성

vsh123 committed:
feat : 로또 생성 기능 구현

vsh123 committed:
refactor : LottoNumber객체 리팩토링

vsh123 committed:
feat : 구입한 금액에 맞게 로또를 생성하는 기능 구현

vsh123 committed:
feat : 구입금액을 입력받고 로또생성하여 출력하는 기능 구현

vsh123 committed:
feat : 당첨 번호를 입력받는 기능 구현

vsh123 committed:
feat : 당첨 결과를 알려주는 Enum객체 생성

vsh123 committed:
feat : 당첨 결과를 저장하고 있는 객체 생성

vsh123 committed:
feat : 당첨 통계를 알려주는 기능 구현

vsh123 committed:
docs : 2단계 기능 구현 목록 작성

vsh123 committed:
feat : 보너스 볼을 입력받아 저장하는 기능 구현

vsh123 committed:
refactor : 메소드 이름 변경

vsh123 committed:
feat : 2등 계산 기능 및 출력 기능 추가

vsh123 committed:
docs : 3단계 기능 구현 목록 작성

vsh123 committed:
refactor : 수동과 자동을 만들어주는 Generator 분리

vsh123 committed:
refactor : WinningLotto 클래스 리팩토링

vsh123 committed:
refactor : LottoResult 리팩토링

vsh123 committed:
feat : 금액을 저장 및 검증하는 Money 클래스 생성

vsh123 committed:
feat : 수동 로또를 입력 받아 생성하는 기능 구현

vsh123 committed:
refactor : 일부 출력 메세지 리팩토링

vsh123 committed:
refactor : Main Controller 리팩토링

 1. void 메소드에 대한 재귀 호출 변경

vsh123 committed:
refactor : if문 분기 조건 변경

woowahan-pjs commented on src/main/java/lottogame/LottoGameMain.java:
```diff
@@ -0,0 +1,93 @@
+    private static Money getMoney() {
+        try {
+            return new Money(Integer.parseInt(InputView.getPrice()));
+        } catch (NumberFormatException | InvalidLottoPriceException e) {
```
캡틴께서 "반복문 대신 재귀 함수로 구현할 수 있다."라고 알려주셨어요. 하지만 재귀를 사용하면 어떤 위험이 있을까요? 만약 재귀를 사용하지 않는다면 어떻게 문제를 해결할 수 있을까요?

vsh123 replied:
재귀를 사용한다면 메모리에 대한 오버플로우가 발생할 위험이 있습니다. 이를 해결하기 위해 반복문으로 변경을 하였고 try-catch를 각 도메인에서 실행하였습니다!

woowahan-pjs commented on src/main/java/lottogame/view/InputView.java:
```diff
@@ -0,0 +1,31 @@
+public class InputView {
+    private static final Scanner SCANNER = new Scanner(System.in);
+
+    public static String getPrice() {
```
Controller에서 구입 금액을 검증할 수 있도록 int가 아닌 String을 반환하네요! 💯

woowahan-pjs commented on src/main/java/lottogame/utils/InvalidLottoPriceException.java:
```diff
@@ -0,0 +1,7 @@
+package lottogame.utils;
+
+public class InvalidLottoPriceException extends RuntimeException {
```
커스터마이징 한 예외 클래스 👍

woowahan-pjs commented on src/main/java/lottogame/utils/InvalidLottoPriceException.java:
```diff
@@ -0,0 +1,7 @@
+package lottogame.utils;
```
`InvalidLottoPriceException`은 `utils` 패키지에 있는 것이 적절할까요?

woowahan-pjs commented on src/main/java/lottogame/domain/Money.java:
```diff
@@ -0,0 +1,34 @@
+        this.price = price;
+    }
+
+    private boolean isValidatePrice(int price) {
```
여기서 `InvalidLottoPriceException`이 발생하면 어떨까요?

woowahan-pjs commented on src/main/java/lottogame/LottoGameMain.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    private static void createLottos(LottoTickets lottoTickets, Money money) {
+        if (!lottoTickets.isCreatedNumberOfManualLotto()) {
```
아래의 글을 읽고 코드에 반영하면 어떨까요?
https://schneide.blog/2014/08/03/dont-ever-not-avoid-negative-logic

woowahan-pjs commented on src/main/java/lottogame/view/OutputView.java:
```diff
@@ -0,0 +1,24 @@
+public class OutputView {
+    public static void printPurchaseResult(LottoTickets lottoTickets) {
+        System.out.println("수동으로 " + lottoTickets.getNumberOfManualTicket() + "개, 자동으로 " +
+                (lottoTickets.numberOfLottos() - lottoTickets.getNumberOfManualTicket()) + "개를 구매했습니다.");
```
`lottoTickets.numberOfLottos() - lottoTickets.getNumberOfManualTicket()`은 의미를 설명할 수 있는 메서드로 한 번 더 분리하는 것이 좋아 보여요.

woowahan-pjs commented on src/main/java/lottogame/view/OutputView.java:
```diff
@@ -0,0 +1,24 @@
+        System.out.println("당첨 통계");
+        System.out.println("---------");
+        System.out.println(lottoResult);
+        System.out.println("총 수익률은 " + lottoResult.getRateOfLotto(money) + "%입니다.");
```
View는 Model의 주요 기능을 실행하는 것이 아닌 화면을 보여주는 데에 집중하면 어떨까요?

woowahan-pjs commented on src/main/java/lottogame/domain/Money.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    long rateOf(double profits) {
+        return (long) (profits / price) * PERCENTAGE;
```
아래의 글은 가볍게 참고만 해주세요.
https://isooo.github.io/java/2019/01/17/BigDecimal.html

woowahan-pjs commented on src/main/java/lottogame/domain/LottoResult.java:
```diff
@@ -0,0 +1,31 @@
+    public String toString() {
+        StringBuilder stringBuilder = new StringBuilder();
+        for (Rank rank : result.keySet()) {
+            stringBuilder.append(rank).append(result.get(rank)).append("개\n");
```
Model에 View의 요구 사항을 반영하는 것이 맞을까요? Model과 View의 기능을 분리하면 어떤 장점이 있을까요?

vsh123 replied:
Model과 View의 기능을 분리하게 된다면 다양한 View에 맞게 유동적으로 보여줄 수 있다는 장점이 있습니다!

woowahan-pjs commented on src/main/java/lottogame/domain/WinningLotto.java:
```diff
@@ -0,0 +1,34 @@
+
+    public WinningLotto(List<Integer> winningLotto) {
+        this.winningLotto = ManualLottoGenerator.create(winningLotto);
+        this.bonusNumber = null;
```
`WinningLotto`는 보너스 번호 없이 존재할 수 있을까요? 도메인 객체가 불완전한 상태로 사용되는 것을 막으려면 생성 시점에 필요한 것을 전달해 주어야 합니다.

woowahan-pjs commented on src/main/java/lottogame/domain/Rank.java:
```diff
@@ -0,0 +1,52 @@
+    @Override
+    public String toString() {
+        if (this == SECOND) {
+            return numberOfMatch + "개 일치, 보너스 볼 일치(" + prize + "원) - ";
```
마찬가지로 Model에 View의 요구 사항을 반영하는 것이 맞을까요?

woowahan-pjs commented on src/main/java/lottogame/domain/Rank.java:
```diff
@@ -0,0 +1,52 @@
+    FOURTH(4, 50000),
+    THIRD(5, 1500000),
+    SECOND(5, 30000000),
+    FIRST(6, 2000000000);
```
Java7부터는 `100_000_000`처럼 숫자에 밑줄을 사용하여 가독성을 높일 수 있어요.

woowahan-pjs commented on src/main/java/lottogame/domain/LottoTickets.java:
```diff
@@ -0,0 +1,61 @@
+
+public class LottoTickets {
+    private List<Lotto> lottos = new ArrayList<>();
+    private final int numberOfManualTicket;
```
`LottoTickets`에서 `numberOfManualTicket`을 관리하는 것이 맞을까요? 아래의 글을 읽고 일급 컬렉션 만들기에 도전해보세요!
https://jojoldu.tistory.com/412

woowahan-pjs commented on src/main/java/lottogame/domain/ManualLottoGenerator.java:
```diff
@@ -0,0 +1,24 @@
+import java.util.Set;
+
+public class ManualLottoGenerator {
+    private static final int LOTTO_SIZE = 6;
```
`AutoLottoGenerator`와 `ManualLottoGenerator`에서 코드 중복이 발생하였네요. `LOTTO_SIZE`를 적절한 위치로 이동시켜 보세요.

woowahan-pjs commented on src/main/java/lottogame/domain/LottoResultGenerator.java:
```diff
@@ -0,0 +1,23 @@
+        }
+    }
+
+    public static LottoResult create(LottoTickets lottoTickets, WinningLotto winningLotto) {
```
이 메서드는 두 번 이상 사용할 수 있을까요?

woowahan-pjs commented on src/main/java/lottogame/domain/LottoNumber.java:
```diff
@@ -0,0 +1,63 @@
+        for (int lottoNumber = MIN_LOTTO_NUMBER; lottoNumber <= MAX_LOTTO_NUMBER; lottoNumber++) {
+            lottoNumbers.put(lottoNumber, new LottoNumber(lottoNumber));
+        }
+    }
```
인스턴스를 미리 만들어 놓거나 새로 생성한 인스턴스를 캐싱 하여 재활용하는 식으로 불필요한 객체 생성을 피할 수 있다. 👍

woowahan-pjs commented on src/main/java/lottogame/domain/WinningLotto.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    boolean isBonusBallMatch(LottoNumber lottoNumber) {
+        return bonusNumber == lottoNumber;
```
인스턴스를 캐싱 하여 재활용하였기에 사용 가능한 코드 👍

woowahan-pjs reviewed (changes requested):
전체적인 설계와 깔끔한 코드 구현이 좋았습니다. 💯 
몇 가지 의견을 남겨놓았으니 충분히 고민하고 도전해보세요!
진행 과정에서 궁금한 점은 댓글이나 Slack을 통해 물어보세요.

vsh123 committed:
refactor : exception 관련 피드백 반영

 1. exception 클래스 패키지 변경
 2. Exception 발생 위치 변경

vsh123 committed:
refactor : 조건문 피드백 반영

 1. not(!)을 붙이는 것 대신 조건문의 내용을 바꿈

vsh123 committed:
refactor : Controller 재귀 로직 변경

 1. 기존 재귀함수에서 반복문으로 변경
 2. 예외에 대한 모든 catch는 도메인에서 실시

vsh123 committed:
refactor : View 및 LottoGenerator 피드백 반영

 1. Model에서 View를 만들어주는 것이 아닌 필요한 데이터를 넘기는 방식으로 변경
 2. Auto와 Manual Generator를 하나의 클래스로 합침

vsh123 committed:
refactor : LottoResultGenerator 삭제

 1. LottoResultGenerator는 한번만 사용 되기 때문에 별도로 빼지 않고
    LottoResult클래스 안의 generate메소드로 대체

vsh123 committed:
refactor : LottoTickets 객체 수정

 1. LottoTickets 객체를 일급 컬렉션으로 변경

vsh123 commented:
안녕하세요! 보내주신 피드백 반영하여 1차 수정 진행하였습니다.
혹시 더 수정해야할 부분이 있다면 언제든지 피드백 부탁드리겠습니다!

항상 감사합니다~

woowahan-pjs commented on src/main/java/lottogame/domain/LottoTickets.java:
```diff
@@ -0,0 +1,52 @@
+
+    public void addManualLotto(String manualLotto) {
+        Lotto lotto = LottoGenerator.create(manualLotto);
+        if(lotto != null){
```
IntelliJ IDEA의 코드 자동 정렬 기능(⌥⌘L, Ctrl+Alt+L)을 사용하면 더 깔끔한 코드를 볼 수 있을 거예요.

woowahan-pjs commented on src/main/java/lottogame/domain/Money.java:
```diff
@@ -0,0 +1,53 @@
+        try {
+            return getNumberOfTicket() < Integer.parseInt(numberOfManualLotto);
+        } catch (NumberFormatException e) {
+            System.out.println("숫자를 입력해 주세요.");
```
View의 요구 사항이네요.

woowahan-pjs commented on src/main/java/lottogame/view/InputView.java:
```diff
@@ -0,0 +1,39 @@
+        do{
+            System.out.println("수동으로 구매할 로또 수를 입력해 주세요.");
+            input = SCANNER.nextLine();
+        } while(money.isInvalidNumber(input));
```
View는 Model의 주요 기능을 실행하는 것이 아닌 화면을 보여주는 데에 집중하면 어떨까요?

woowahan-pjs commented on src/main/java/lottogame/view/OutputView.java:
```diff
@@ -0,0 +1,42 @@
+        for (Rank rank : Rank.values()) {
+            printMatchResultOf(rank, lottoResult);
+        }
+        System.out.println("총 수익률은 " + rateOfLotto + "%입니다.");
```
View는 Model의 주요 기능을 실행하는 것이 아닌 화면을 보여주는 데에 집중한다. 👍

woowahan-pjs commented on src/main/java/lottogame/domain/LottoGenerator.java:
```diff
@@ -0,0 +1,52 @@
+import java.util.Set;
+
+public class LottoGenerator {
+    private static final int LOTTO_SIZE = 6;
```
`ManualLottoGenerator`와 `AutoLottoGenerator`를 `LottoGenerator`로 합친 이유가 있을까요? 테스트 가능한 부분과 테스트하기 힘든 부분은 분리하는 것이 좋아요.

vsh123 replied:
LOTTO_SIZE에 대한 중복으로 인해 합치게 되었습니다.
하지만 아래 comment 달아주신 것을 참고하면 해당 부분을 다시 분리하는 것이 더 좋을 것 같다는 생각이 듭니다!
감사합니다 :)

woowahan-pjs commented on src/test/java/lottogame/domain/LottoGeneratorTest.java:
```diff
@@ -0,0 +1,15 @@
+public class LottoGeneratorTest {
+    @Test
+    void 생성_테스트() {
+        assertThat(LottoGenerator.create("1,2,3,4,5,6")).isExactlyInstanceOf(Lotto.class);
```
조금 더 명확한 테스트 코드를 만들어 보세요.

woowahan-pjs commented on src/main/java/lottogame/domain/LottoGenerator.java:
```diff
@@ -0,0 +1,52 @@
+            return new Lotto(new ArrayList<>(lottoNumbers));
+        } catch (NumberFormatException | InvalidLottoNumberException e) {
+            System.out.println(e.getMessage());
+            return null;
```
정적 팩토리 메서드는 생성자의 기능을 대신합니다. `null`을 반환하는 것이 좋을지 같이 고민해봐요.

woowahan-pjs commented on src/main/java/lottogame/domain/LottoGenerator.java:
```diff
@@ -0,0 +1,52 @@
+import java.util.Set;
+
+public class LottoGenerator {
+    private static final int LOTTO_SIZE = 6;
```
`LOTTO_SIZE`는 `LottoGenerator`가 아닌 `Lotto`에 위치할 수도 있겠죠? 첨언을 하면 public static final 필드 방식을 사용해 보세요.

woowahan-pjs commented on src/main/java/lottogame/domain/LottoResult.java:
```diff
@@ -0,0 +1,41 @@
+public class LottoResult {
+    private Map<Rank, Integer> result = new LinkedHashMap<>();
+
+    LottoResult() {
```
`LottoResult`를 생성할 때 완전한 상태로 생성할 수 있게 변경하면 어떤 장점이 있을까요?

woowahan-pjs commented on src/main/java/lottogame/domain/LottoResult.java:
```diff
@@ -0,0 +1,41 @@
+        result.remove(Rank.MISS);
+    }
+
+    public static LottoResult generate(LottoTickets lottoTickets, WinningLotto winningLotto) {
```
`LottoResult`가 `LottoTickets`와 `WinningLotto`에 의존하고 있어요. 삭제한 `LottoResultGenerator`가 필요할 것 같아요. 아래의 코멘트는 기존 `LottoResultGenerator`에서 `create()`를 두 번 이상 호출하면 예상치 못한 값을 가진 `LottoResult`가 생성된다는 뜻이었습니다.
https://github.com/woowacourse/java-lotto/pull/11#discussion_r291832974

woowahan-pjs commented on src/main/java/lottogame/domain/LottoResult.java:
```diff
@@ -0,0 +1,41 @@
+        return lottoResult;
+    }
+
+    public long getRateOfLotto(Money money) {
```
테스트 코드가 필요해 보여요.

woowahan-pjs commented on src/main/java/lottogame/lottogameexception/InvalidLottoNumberException.java:
```diff
@@ -0,0 +1,7 @@
+package lottogame.lottogameexception;
```
`InvalidLottoPriceException`이 `domain` 패키지에 위치하면 어떨까요? 같이 고민해봐요.

woowahan-pjs reviewed (changes requested):
테스트 코드를 꾸준히 작성하면 더 깔끔한 코드를 볼 수 있을 거예요.
몇 가지 의견을 남겨놓았으니 충분히 고민하고 도전해보세요!
다음 리뷰 요청 때는 Merge 하도록 하겠습니다.

vsh123 committed:
refactor : Model,View 역할 분리

 1. View와 Model의 역할을 좀 더 명확하게 분리
 2. 자바 컨벤션 준수

vsh123 committed:
refactor : LottoGenerator 분리

 1. Auto와 Manual을 분리
 2. Manual에 대한 테스트 코드 작성

vsh123 committed:
refactor : 정적 팩토리 메소드 리팩토링

 1. 정적 팩토리 메소드에서 null을 리턴하지 않고
    정확한 값이 들어갔는지 확인하는 로직을 추가

vsh123 committed:
refactor : 테스트 코드 작성 및 Exception 위치 변경

vsh123 committed:
refactor : LottoResult 리팩토링

 1. LottoResultGenerator 생성

vsh123 commented:
안녕하세요! 보내주신 피드백 반영하여 다시한번 리뷰 요청드립니다.

이번 피드백 리뷰를 반영하면서 정적 팩토리 메서드에서 null을 반환하는 것에 대해 많은 고민이 있었습니다ㅠㅠ
코드 보시고 추가적으로 고칠 부분이 보인다면 코멘트 부탁드리겠습니다!!

감사합니다!

vsh123 committed:
refactor : try-catch 및 View에 대한 반복 로직 변경

 1. 기존 Model에서 실시하던 try-catch를 controller에서 실시
 2. Optional을 사용하여 반복문을 실행

woowahan-pjs commented on src/main/java/lottogame/LottoGameMain.java:
```diff
@@ -0,0 +1,112 @@
+            System.out.println(e.getMessage());
+            return null;
+        }
+    }
```
아래의 코드로 개선해보면 어떨까요?

```java
private static Optional<Money> createMoney(String price) {
    try {
        return Optional.of(Money.generate(Integer.parseInt(price)));
    } catch (RuntimeException e) {
        System.out.println(e.getMessage());
        return Optional.empty();
    }
}
```

woowahan-pjs commented on src/main/java/lottogame/domain/LottoTickets.java:
```diff
@@ -0,0 +1,48 @@
+public class LottoTickets {
+    private List<Lotto> lottos = new ArrayList<>();
+
+    public static LottoTickets generate() {
```
`LottoTickets`를 생성할 때 완전한 상태로 생성할 수 있게 변경하면 어떤 장점이 있을까요?

woowahan-pjs reviewed (approved):
피드백을 모두 잘 반영해주셨습니다. 💯 
생성자 대신 사용하는 것이 정적 팩토리 메서드입니다.
다른 개발자가 정적 팩토리 메서드를 사용할 때 `null`이 반환될 것을 예상할 수 있을까요?
고민에 대한 답으로 아래의 글이 도움 되면 좋겠어요.

- https://www.codebyamir.com/blog/stop-returning-null-in-java
- https://stackoverflow.com/questions/11124562/is-it-ok-for-a-factory-method-to-return-null

다음 단계를 진행하려면 woowacouse에 Merge 된 코드에서 다시 시작해야 충돌이 발생하지 않을 거예요.
아래의 과정을 진행한 후, 다음 단계 진행할 것을 추천해요.

```
1. git remote add upstream
2. git fetch upstream 자신의 github_id
3. git reset --hard upstream/자신의 github_id
```

다음 단계에서 뵙겠습니다. 이만 Merge 하겠습니다.
